new: Show mod version on F3 screen

Leo40Git opened the pull request:
Shows "(!!) Dev Environment" in red if we're in the mod's development environment (IE if the version is the ${version} placeholder).
**Note:** This preview is outdated! Scroll down for an up-to-date preview.
![Preview](https://media.discordapp.net/attachments/651120262129123330/823132648473362443/unknown.png)

Leo40Git committed:
new: Show mod version on F3 screen

Show "(!!) Devel" in red if we're in the mod's development environment (IE if the version is the ${version} placeholder)

Leo40Git committed:
oops: remove unused import

CalXee commented:
May I ask what "!! Devel" means?

ImperatorStorm commented:
> May I ask what "!! Devel" means?

commit [f55422b](https://github.com/CaffeineMC/sodium-fabric/pull/598/commits/f55422bcef6520674889f3af424cd36436e51d55)
> Show "(!!) Devel" in red if we're in the mod's development environment (IE if the version is the ${version} placeholder)

DragonEggBedrockBreaking commented:
> May I ask what "!! Devel" means?

"(!!)" is likely an exclamation/a small warning, and "Devel" is short for development environment

machinesmith42 commented:
would deven be right then?

silversquirl commented on src/main/java/me/jellysquid/mods/sodium/client/SodiumClientMod.java:
```diff
@@ -17,6 +21,18 @@ public void onInitializeClient() {
+                    .orElseThrow(() -> new IllegalStateException("Mod with ID \"sodium\" couldn't be found... but that's us!"));
+            VERSION = ourContainer.getMetadata().getVersion().getFriendlyString();
+            if ("${version}".equals(VERSION))
+                VERSION = Formatting.RED + "(!!) Devel";
```
For clarity, and in order to match the style elsewhere in Sodium, this should be either on the same line as the if:

```java
            if ("${version}".equals(VERSION)) VERSION = Formatting.RED + "(!!) Devel";
```

Or wrapped in braces:

```java
            if ("${version}".equals(VERSION)) {
                VERSION = Formatting.RED + "(!!) Devel";
            }
```

I would recommend the latter.

silversquirl commented on src/main/java/me/jellysquid/mods/sodium/mixin/features/debug/MixinDebugHud.java:
```diff
@@ -35,6 +36,9 @@ private void appendRightText(CallbackInfoReturnable<List<String>> cir) {
+        strings.add("");
+        strings.add("Sodium Version: " + SodiumClientMod.version());
+
         strings.add("");
```
I think it would make sense to remove the spacing here, so the two things added by Sodium are grouped

Leo40Git committed:
Move version string onto its own line

Since long version strings are possible (rev.<commit_id>), this makes sure that the version is always visible
Extended "(!!) Devel" to "(!!) Dev Environment" as well

Leo40Git commented:
Updated preview:
![image](https://user-images.githubusercontent.com/7997354/112110772-0a6d9b80-8bbc-11eb-877f-0a7e624a5699.png)

Boobies commented:
Is this really necessary? If every mod did this, the debug screen would be a disaster. Presumably, the user who installed the mod, knows what version they are running and, if not, they can easily check their mods directory. I see this PR as mildly detrimental at best.

vktec commented:
Knowing the version is incredibly helpful for dealing with support queries when many people are running builds of the next branch or forks of the project.

DragonEggBedrockBreaking commented:
> Knowing the version is incredibly helpful for dealing with support queries when many people are running builds of the next branch or forks of the project.

Just to add on to what vktec said, people do not usually run development builds of most other mods (in my experience), so there is very little reason for "every mod" to do this.

Meeni-mc commented:
Maybe it would make sense to have the git commit hash rather than a generic 'Dev'.

Also having more frequent releases (publish anything that is an improvement in A.B, do not hold until feature X is finally implemented, I't ll go in A.B+1 just fine) may help with people using Dev versions left and right.

Leo40Git commented:
> Maybe it would make sense to have the git commit hash rather than a generic 'Dev'.

The generic "Dev" is only used if someone's playing around in Sodium's development environment (_not_ in any development environment).

Boobies commented:
> Knowing the version is incredibly helpful for dealing with support queries when many people are running builds of the next branch or forks of the project.

Knowing that stuff is useful but I don't see what that has to do with putting it on the debug screen. Could also print it on the hotbar and under each button in the main menu since knowing the version is useful. Or, when people are asking for support, they could say what instead they're instead of posting F3 screenshots.

vktec commented:
We ask for F3 screenshots anyway since they provide other useful information. It's called the debug screen for a reason, after all :)
